sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Reify.java

```diff
@@ -0,0 +1,140 @@
+ */
+public class Reify {
+  /**
+   * Reify information from the processing context into an instance of {@link ValueInSingleWindow}.
+   *
+   * @param <T> element type
+   */
+  private static class ValuesInWindowFn<T> extends DoFn<T, ValueInSingleWindow<T>> {
+    @ProcessElement
+    public void processElement(ProcessContext c, BoundedWindow window) {
+      c.outputWithTimestamp(ValueInSingleWindow.of(c.element(), c.timestamp(), window, c.pane()),
+          c.timestamp());
+    }
+  }
+
+  private static class Window<T> extends PTransform<PCollection<T>,
+      PCollection<ValueInSingleWindow<T>>> {
+    @Override
+    public PCollection<ValueInSingleWindow<T>> expand(PCollection<T> input) {
+      return input
+          .apply(ParDo.of(new ValuesInWindowFn<T>()))
+          .setCoder(ValueInSingleWindow.Coder.of(input.getCoder(),
+              input.getWindowingStrategy().getWindowFn().windowCoder()));
+    }
+  }
+
+  private static class WindowInValue<K, V> extends PTransform<PCollection<? extends KV<K, V>>,
+      PCollection<KV<K, ValueInSingleWindow<V>>>> {
+    @Override
+    public PCollection<KV<K, ValueInSingleWindow<V>>> expand(PCollection<? extends KV<K, V>>
+        input) {
+      List<? extends Coder<?>> coderArguments = input.getCoder().getCoderArguments();
+      Coder<K> keyCoder = (Coder<K>) coderArguments.get(0);
+      Coder<V> valueCoder = (Coder<V>) coderArguments.get(1);
+      return input
+          .apply(ParDo.of(new WindowInValueFn<K, V>()))
+          .setCoder(KvCoder.of(keyCoder, ValueInSingleWindow.Coder.of(valueCoder,
+              input.getWindowingStrategy().getWindowFn().windowCoder())));
+    }
+
+    private class WindowInValueFn<K, V> extends DoFn<KV<K, V>, KV<K, ValueInSingleWindow<V>>> {
+      @ProcessElement
+      public void processElement(ProcessContext c, BoundedWindow window) {
+        c.output(KV.of(c.element().getKey(), ValueInSingleWindow.of(c.element()
+            .getValue(), c.timestamp(), window, c.pane())));
+      }
+    }
+  }
+
+  private static class ReifyValueTimestampDoFn<K, V>
+      extends DoFn<KV<K, V>, KV<K, TimestampedValue<V>>> {
+    @ProcessElement
+    public void processElement(ProcessContext context) {
+      context.output(
+          KV.of(
+              context.element().getKey(),
+              TimestampedValue.of(context.element().getValue(), context.timestamp())));
+    }
+  }
+
+  private static class ReifyTimestampDoFn<T> extends DoFn<T, TimestampedValue<T>> {
+    @ProcessElement
+    public void processElement(ProcessContext context) {
+      context.output(TimestampedValue.of(context.element(), context.timestamp()));
+    }
+  }
+
+  Reify() {
+  }
+
+  /**
+   * Create a {@link PTransform} that will output all input {@link KV KVs} with the timestamp inside
+   * the value.
+   */
+  public static <K, V>
+  PTransform<PCollection<? extends KV<K, V>>, PCollection<KV<K, TimestampedValue<V>>>>
```

Comment:
Please name this one consistently with the other one, windowsInValue - e.g. timestampsInValue.

Also, I suggest to drop the "? extends" throughout the PR - this wildcard stuff rarely makes the transforms more usable (especially because KV is not intended to be extended, and generally classes used as PCollection elements are rarely parts of an inheritance hierarchy), but often leads to more fighting with the type checker.